hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -399,4 +795,119 @@ public static int mapRecordKeyToFileGroupIndex(String recordKey, int numFileGrou
+      throw new HoodieException("Failed to generate column stats records for metadata table ", e);
+    }
+  }
+
+  /**
+   * Create column stats from write status.
+   *
+   * @param engineContext                       - Enging context
+   * @param datasetMetaClient                   - Dataset meta client
+   * @param allWriteStats                       - Write status to convert
+   * @param isMetaIndexColumnStatsForAllColumns - Are all columns enabled for indexing
+   */
+  public static List<HoodieRecord> createColumnStatsFromWriteStats(HoodieEngineContext engineContext,
+                                                                   HoodieTableMetaClient datasetMetaClient,
+                                                                   List<HoodieWriteStat> allWriteStats,
+                                                                   boolean isMetaIndexColumnStatsForAllColumns) throws Exception {
+    if (allWriteStats.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    List<HoodieWriteStat> prunedWriteStats = allWriteStats.stream().filter(writeStat -> {
+      return !(writeStat instanceof HoodieDeltaWriteStat);
+    }).collect(Collectors.toList());
+    if (prunedWriteStats.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    return engineContext.flatMap(prunedWriteStats,
+        writeStat -> translateWriteStatToColumnStats(writeStat, datasetMetaClient,
+            getLatestColumns(datasetMetaClient, isMetaIndexColumnStatsForAllColumns)),
+        prunedWriteStats.size());
+  }
+
+  /**
+   * Get the latest columns for the table for column stats indexing.
+   *
+   * @param datasetMetaClient                   - Data table meta client
+   * @param isMetaIndexColumnStatsForAllColumns - Is column stats indexing enabled for all columns
+   */
+  private static List<String> getLatestColumns(HoodieTableMetaClient datasetMetaClient, boolean isMetaIndexColumnStatsForAllColumns) {
+    if (!isMetaIndexColumnStatsForAllColumns) {
+      return Collections.singletonList(datasetMetaClient.getTableConfig().getRecordKeyFieldProp());
+    }
+
+    if (datasetMetaClient.getCommitsTimeline().filterCompletedInstants().countInstants() > 1) {
+      TableSchemaResolver schemaResolver = new TableSchemaResolver(datasetMetaClient);
+      // consider nested fields as well. if column stats is enabled only for a subset of columns,
+      // directly use them instead of all columns from the latest table schema
+      try {
+        return schemaResolver.getTableAvroSchema().getFields().stream()
+            .map(entry -> entry.name()).collect(Collectors.toList());
+      } catch (Exception e) {
+        throw new HoodieException("Failed to get latest columns for " + datasetMetaClient.getBasePath());
+      }
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  private static List<String> getLatestColumns(HoodieTableMetaClient datasetMetaClient) {
+    return getLatestColumns(datasetMetaClient, false);
```

Comment:
so, looks like we index only record key in this patch is it ?
I mean, we don't support adding index for all cols. is my understanding right.

Reply:
Right, HUDI-3260 tracks multiple columns.